src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,771 @@
+            this.logger.LogTrace("({0}:{1},{2}:{3},{4}:{5})", nameof(downloadJob.Hashes.Count), downloadJob.Hashes.Count, nameof(failedJobs.Count), failedJobs.Count, nameof(emptySlotes), emptySlotes);
+
+            var newAssignments = new Dictionary<uint256, AssignedDownload>();
+            
+            var peers = new Dictionary<int, ChainedHeader>(this.peerIdsToTips);
+            bool jobFailed = false;
+
+            foreach (uint256 hashToAssign in downloadJob.Hashes.Take(emptySlotes).ToList())
+            {
+                ChainedHeader header = this.HeadersByHash[hashToAssign];
+
+                while (!jobFailed)
+                {
+                    double sumOfQualityScores = this.pullerBehaviorsByPeerId.Values.Sum(x => x.QualityScore);
+
+                    double scoreToReachPeer = this.random.Next(0, (int)(sumOfQualityScores * 1000)) / 1000.0;
+
+                    int peerId = 0;
+
+                    foreach (BlockPullerBehavior pullerBehavior in this.pullerBehaviorsByPeerId.Values)
+                    {
+                        if (pullerBehavior.QualityScore >= scoreToReachPeer)
+                            peerId = pullerBehavior.AttachedPeer.Connection.Id;
+                        else
+                            scoreToReachPeer -= pullerBehavior.QualityScore;
+                    }
+
+                    ChainedHeader peerTip = peers[peerId];
+
+                    if (peerTip.GetAncestor(header.Height) == header)
+                    {
+                        // Assign to this peer
+                        newAssignments.Add(hashToAssign, new AssignedDownload()
+                        {
+                            PeerId = peerId,
+                            JobId = downloadJob.Id,
+                            AssignedTime = DateTime.UtcNow,
+                            BlockHeight = header.Height
+                        });
+
+                        this.logger.LogTrace("Block '{0}' was assigned to peer {1}", hashToAssign, peerTip);
+
+                        downloadJob.Hashes.Remove(hashToAssign);
+                        break;
+                    }
+                    else
+                    {
+                        // Peer doesn't claim this hash.
+                        peers.Remove(peerId);
+
+                        if (peers.Count != 0)
+                            continue;
+
+                        jobFailed = true;
+
+                        this.logger.LogDebug("Job {0} failed because there is no peer claiming {1} of it's hashes.", downloadJob.Id, downloadJob.Hashes.Count);
+                    }
+                }
+
+                this.pendingDownloadsCount += newAssignments.Count;
+            }
+
+            if (jobFailed)
+            {
+                failedJobs.Add(new DownloadJob() {Hashes = downloadJob.Hashes });
+
+                downloadJob.Hashes = new HashSet<uint256>();
+            }
+
+            this.logger.LogTrace("(-).*:{0}", newAssignments.Count);
```

Comment:
`this.logger.LogTrace("(-):*.{0}:{1}", nameof(newAssignments.Count), newAssignments.Count);`

Reply:
Fixed